src/Microsoft.Azure.WebJobs.Host/Retry/RetryManager.cs

```diff
@@ -0,0 +1,159 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT License. See License.txt in the project root for license information.
+
+using System;
+using System.Threading.Tasks;
+using Microsoft.Extensions.Logging;
+using Polly;
+using Polly.Retry;
+using Polly.Fallback;
+using System.Threading;
+using Microsoft.Azure.WebJobs.Host.Executors;
+using System.Collections.Generic;
+
+namespace Microsoft.Azure.WebJobs.Host
+{
+    /// <summary>
+    /// The class that implements a function execution with retries.
+    /// </summary>
+    public class RetryManager
+    {
+        private readonly ILogger _logger;
+        private readonly string _id;
+        private RetryResult _retry;
+
+        public RetryManager(RetryPolicyOptions retryPolicyOptions, ILogger logger)
+        {
+            _logger = logger;
+            _id = Guid.NewGuid().ToString();
+
+            if (retryPolicyOptions != null && retryPolicyOptions.IsDefined)
+            {
+                _retry = retryPolicyOptions as RetryResult;
+            }
+        }
+
+        /// <summary>
+        /// Executes a specified function with retries.
+        /// </summary>
+        /// <param name="execute">The action to perform.</param>
+        /// <param name="executeParams">Inpput arameters for the action.</param>
+        /// <param name="token">A cancellation token which can be used to cancel the action. When a retry policy
+        /// in use, also cancels any further retries.</param>
+        public async Task ExecuteWithRetriesAsync(Func<int, object[], Task<IEnumerable<FunctionResult>>> execute, object[] executeParams, CancellationToken token)
```

Comment:
I don't see where this is being called? I'd expect to see FunctionExecutor have an "execute with retries" function, similar to the existing "execute with timeout", and handle failures/retries there. As it stands, it looks like this isn't yet a general function retry mechanism? E.g. Be able to apply a [RetryPolicy] attribute to a function, with corresponding support in Azure Functions function.json. That's what I expected when I saw your PR title.

However, maybe you're not trying to provide that general feature for invocation retries, but instead something targeted only for EH checkpointing retries?  Then I'd expect this core SDK to just expose some general RetryPolicy types, and the EH extension simply uses them, and handles all checkpointing retries internally, without any changes to FunctionExecutor. In the future when we want to add a [RetryPolicy] attribute for function retries, we'd leverage that RetryPolicy type.

Reply:
The calls:
https://github.com/Azure/azure-webjobs-sdk/pull/2463/files#diff-d4edc2f1f55dd03bebe0b9fa640a1656R391 (load RetryManagerTests.cs in the PR)

https://github.com/Azure/azure-functions-eventhubs-extension/pull/56/files#diff-195d2c9ccd88ec324d536317b700333aR204

The idea of `RetryManager` is to execute with retries code which returns `IEnumerable<FunctionResult>` so we can use it for EH/ComsosDB/Kafka listeners. We do not want `[RetryPolicy]` attribute for any trigger.

Design of the feature is here:
https://github.com/jeffhollan/retry-design